server/src/main.c

```diff
@@ -0,0 +1,404 @@
+      continue;
+    }
+
+    // Add thread to linked list
+    slist_data->thread = thread_id;
+    SLIST_INSERT_HEAD(&head, slist_data, nodes);
+  }
+
+  // Wait for all threads to join
+  if (slist_join_threads(&head)) {
+    application_result = -1;
+  }
+
+  slist_free(&head);
+  return application_result;
+}
+
+int setup_socket_server(const bool execute_as_daemon, int *server_socket,
+                        struct addrinfo **server_addrinfo) {
+  // Setup the socket server
+  if (socket_server_create(server_socket, server_addrinfo) != 0) {
+    syslog(LOG_ERR, "create_server_socket");
+    return -1;
+  }
+
+  // Handle daemon execution if selected. Must be done after socket is setup
+  if (execute_as_daemon) {
+    const int daemon_resut = daemonize();
+    switch (daemon_resut) {
+    case -1:
+      syslog(LOG_ERR, "daemonize");
+      return -1;
+    case 1:
+      exit(EXIT_SUCCESS);
+    default:
+    }
+  }
+
+  return 0;
+}
+
+int setup_signal_handlers(timer_t *sugusr1_timer) {
+  // Register SIGINT handler
+  if (signal(SIGINT, sigint_handler) == SIG_ERR) {
+    perror("signal");
+    return -1;
+  }
+
+  // Register SIGTERM handler
+  if (signal(SIGTERM, sigint_handler) == SIG_ERR) {
+    perror("signal");
+    return -1;
+  }
+
+  if (signal(SIGUSR1, sigusr1_handler) == SIG_ERR) {
+    perror("signal");
+    return -1;
+  }
+
+  // Initialize clock in the child (if daemonized) with a 10s loop
+  if (sem_init(config_get_timestamp_semaphore(), 0 /* shared between threads */,
+               0)) {
+    perror("sem_init");
+    return -1;
+  }
+
+  struct sigevent evp;
+  evp.sigev_value.sival_ptr = sugusr1_timer;
+  evp.sigev_notify = SIGEV_SIGNAL;
+  evp.sigev_signo = SIGUSR1;
+
+  if (timer_create(CLOCK_MONOTONIC, &evp, sugusr1_timer)) {
+    perror("timer_create");
+    return -1;
+  }
+
+  struct itimerspec timestamp_log_itimespec;
+  timestamp_log_itimespec.it_interval.tv_sec = TIMESTAMP_LOG_INTERVAL_S;
+  timestamp_log_itimespec.it_interval.tv_nsec = 0;
+  timestamp_log_itimespec.it_value.tv_sec = TIMESTAMP_LOG_INTERVAL_S;
+  timestamp_log_itimespec.it_value.tv_nsec = 0;
+
+  if (timer_settime(*sugusr1_timer, 0, &timestamp_log_itimespec, NULL)) {
+    perror("timer_settime");
+    return -1;
+  }
+
+  return 0;
+}
+
+void *data_transfer_worker(void *arg) {
+  ThreadData *thread_data = (ThreadData *)(arg);
+  thread_data->thread_status = RUNNING;
+  syslog(LOG_DEBUG, "Thread %ld started for client %d.", pthread_self(),
+         thread_data->client_fd);
+
+  pthread_mutex_lock(config_get_result_file_mutex());
+  if (socket_client_receive_and_write_data(RESULT_FILE,
+                                           thread_data->client_fd) == -1) {
+    pthread_mutex_unlock(config_get_result_file_mutex());
+    syslog(LOG_ERR, "receive_data");
+    close(thread_data->client_fd);
+    thread_data->thread_status = FAILED;
+    pthread_exit(NULL);
+  }
+  pthread_mutex_unlock(config_get_result_file_mutex());
```

Comment:
This is unlocked right before a lock because `socket_client_receive_and_write_data()`, along with `socket_cleinet_send_file()`, should be reworked to allow the lock to be used in a more focused manner (then they won't be right next to each other).

Reply:
Yes.  You could also do
```
pthread_mutex_lock(config_get_result_file_mutex());
result = socket_client_receive_and_write_data(RESULT_FILE,
                                           thread_data->client_fd) == -1) 
pthread_mutex_unlock(config_get_result_file_mutex());
if( result == -1 )
{
  // do error handling which doesn't require lock
}

```